Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/DataSynchronizer.swift

```diff
@@ -1648,43 +1640,45 @@ public class DataSynchronizer: NetworkStateDelegate, FatalErrorListener {
      */
     func insertMany(documents: [Document], in namespace: MongoNamespace) throws -> InsertManyResult? {
-        guard var nsConfig: NamespaceSynchronization = self.syncConfig[namespace] else {
+        guard let nsConfig: NamespaceSynchronization = self.syncConfig[namespace] else {
             throw StitchError.clientError(
                 withClientErrorCode: .couldNotLoadSyncInfo)
         }
 
-        var deferredEventEmittingBlock: DeferredBlock? = nil
-        defer { deferredEventEmittingBlock?() }
-
         let lock = nsConfig.nsLock
-        return try lock.write {
-
-        // Remove forbidden fields from the documents before inserting them into the local collection.
-        let docsForStorage = documents.map { DataSynchronizer.sanitizeDocument($0) }
-        guard let result = try localCollection(for: namespace).insertMany(docsForStorage) else {
-            return nil
-        }
+        let result: InsertManyResult? = try lock.write {
 
-        let eventEmitters = try result.insertedIds.compactMap({ (kv) -> (() -> Void)? in
-            guard let documentId = kv.value else {
+            // Remove forbidden fields from the documents before inserting them into the local collection.
+            let docsForStorage = documents.map { DataSynchronizer.sanitizeDocument($0) }
+            guard let result = try localCollection(for: namespace).insertMany(docsForStorage) else {
                 return nil
             }
-            let document = docsForStorage[kv.key]
-            let event = ChangeEvent<Document>.changeEventForLocalInsert(namespace: namespace,
-                                                                        document: document,
-                                                                        writePending: true)
-            let config = try nsConfig.sync(id: documentId)
-            try config.setSomePendingWrites(atTime: logicalT, changeEvent: event)
-            return { self.emitEvent(documentId: documentId, event: event) }
-        })
 
-        deferredEventEmittingBlock = {
-            self.triggerListening(to: namespace)
+            let eventEmitters = try result.insertedIds.compactMap({ (kv) -> (() -> Void)? in
+                guard let documentId = kv.value else {
+                    return nil
+                }
+                let document = docsForStorage[kv.key]
+                let event = ChangeEvent<Document>.changeEventForLocalInsert(namespace: namespace,
+                                                                            document: document,
+                                                                            writePending: true)
+                let config = try nsConfig.sync(id: documentId)
+                try config.setSomePendingWrites(atTime: logicalT, changeEvent: event)
+                return { self.emitEvent(documentId: documentId, event: event) }
+            })
+
             eventEmitters.forEach({$0()})
```

Comment:
[nit] This should be moved to happen after the `triggerListening` call, since right now the events are being emitted inside the lock, which isn't necessary